src/API.jl

```diff
@@ -1189,14 +1232,48 @@ function precompile(ctx::Context, pkgs::Vector{PackageSpec}; internal_call::Bool
         end
     end
 
+    function expand_indirect_dependencies(direct_deps)
+        function visit!(visited, node, all_deps)
+            if node in visited
+                return
+            end
+            push!(visited, node)
+            for dep in get(Set{Base.PkgId}, direct_deps, node)
+                if !(dep in all_deps)
+                    push!(all_deps, dep)
+                    visit!(visited, dep, all_deps)
+                end
+            end
+        end
+
+        indirect_deps = Dict{Base.PkgId, Set{Base.PkgId}}()
+        for package in keys(direct_deps)
+            # Initialize a set to keep track of all dependencies for 'package'
+            all_deps = Set{Base.PkgId}()
+            visited = Set{Base.PkgId}()
+            visit!(visited, package, all_deps)
+            # Update direct_deps with the complete set of dependencies for 'package'
+            indirect_deps[package] = all_deps
+        end
+        return indirect_deps
+    end
+
+    indirect_deps = expand_indirect_dependencies(depsmap)
+
     # this loop must be run after the full depsmap has been populated
-    for (pkg, pkg_exts) in pkg_exts_map
-        # find any packages that depend on the extension(s)'s deps and replace those deps in their deps list with the extension(s),
-        # basically injecting the extension into the precompile order in the graph, to avoid race to precompile extensions
-        for (_pkg, deps) in depsmap # for each manifest dep
-            if !in(_pkg, keys(exts)) && pkg in deps # if not an extension and depends on pkg
-                append!(deps, pkg_exts) # add the package extensions to deps
-                filter!(!isequal(pkg), deps) # remove the pkg from deps
+    for ext in keys(exts)
+        ext_loadable_in_pkg = Dict{Base.PkgId,Bool}()
+        for pkg in keys(depsmap)
+            is_trigger = in(pkg, depsmap[ext])
+            is_extension = in(pkg, keys(exts))
+            has_triggers = issubset(depsmap[ext], indirect_deps[pkg])
+            ext_loadable_in_pkg[pkg] = !is_extension && has_triggers && !is_trigger
```

Comment:
For 1.10, I think we want to delete the `!is_extension` condition
```suggestion
            # In contrast to 1.11+, on 1.10 both "pkg → ext" and "ext → ext" dependency edges
            # are implied based on transitive dependencies.
            #
            # This condition is inconsistent for "ext → ext" edges, leading to dependency
            # cycles on 1.10, but this behavior is intentionally preserved for now to avoid
            # breaking packages that depend on this (bad) implicit behavior.
            #
            # See https://github.com/JuliaLang/julia/issues/56204#issuecomment-2442652997
            # for the improved behavior this was replaced with in 1.11
            ext_loadable_in_pkg[pkg] = has_triggers && !is_trigger
```

Technically the check you have here is OK, but I prefer deleting the `!is_extension` condition and getting rid of the for loop just above this that was computing these "ext → ext" dependency edges separately

Reply:
Took the liberty of pushing this